src/access/AccessControl.cpp

```diff
@@ -115,12 +137,45 @@ CHIP_ERROR AccessControl::Check(const SubjectDescriptor & subjectDescriptor, con
             {
                 NodeId subject = kUndefinedNodeId;
                 ReturnErrorOnFailure(entry.GetSubject(i, subject));
-                if (subject == subjectDescriptor.subjects[0])
+                if (IsOperationalNodeId(subject))
+                {
+                    if (subject == subjectDescriptor.subject)
+                    {
+                        subjectMatched = true;
+                        break;
+                    }
+                }
+                else if (IsGroupId(subject))
+                {
+                    VerifyOrReturnError(authMode == AuthMode::kGroup, CHIP_ERROR_INVALID_ARGUMENT);
+                    if (subject == subjectDescriptor.subject)
+                    {
+                        subjectMatched = true;
+                        break;
+                    }
+                }
+                else if (IsPAKEKeyId(subject))
+                {
+                    VerifyOrReturnError(authMode == AuthMode::kPase, CHIP_ERROR_INVALID_ARGUMENT);
+                    if (subject == subjectDescriptor.subject)
+                    {
+                        subjectMatched = true;
+                        break;
+                    }
+                }
+                else if (IsCASEAuthTag(subject))
+                {
+                    VerifyOrReturnError(authMode == AuthMode::kCase, CHIP_ERROR_INVALID_ARGUMENT);
+                    if (CheckSubjectDescriptorCATsAgainstEntrySubject(subjectDescriptor.cats, subject))
+                    {
+                        subjectMatched = true;
+                        break;
+                    }
+                }
+                else
                 {
-                    subjectMatched = true;
-                    break;
+                    return CHIP_ERROR_INVALID_ARGUMENT;
```

Comment:
I wouldn't say CHIP_ERROR_INVALID_ARGUMENT, more like CHIP_ERROR_INTERNAL (since there should never be such a case hit, stored data should match one of the cases).

Reply:
Same goes for the VerifyOrReturnError calls above. We will not be allowing changes to create/update an entry with such invalid contents (e.g. subject type not matching auth mode).

Reply:
updated